hub/core/chunk_engine.py

```diff
@@ -385,34 +390,19 @@ def extend(self, samples):
                     samples = samples[1:]
                 if len(samples) > 0:
                     current_chunk = self._create_new_chunk()
-                    updated_chunks.add(current_chunk)
             else:
                 num = int(num_samples_added)
                 enc.register_samples(num)
                 samples = samples[num:]
 
         self.commit_diff.add_data(indexes_added)
+        self.cache.autoflush = initial_autoflush
+        self.cache.maybe_flush()
 
-        for chunk in updated_chunks:
-            self.cache[chunk.key] = chunk  # type: ignore
-        self._write_finalization()
-
-    def _synchronize_cache(self, chunk_keys: List[str] = None):
-        """Synchronizes cachables with the cache.
-
-        Args:
-            chunk_keys (List[str]): List of chunk keys to be synchronized. If None, only the last chunk will be synchronized. Defaults to None.
-        """
-
-        # TODO implement tests for cache size compute
-        # TODO: optimize this by storing all of these keys in the chunk engine's state (posixpath.joins are pretty slow)
-
-        # synchronize chunks
-        if chunk_keys is None:
-            chunk_keys = [self.last_chunk_key]
-        for chunk_key in chunk_keys:
-            chunk = self.get_chunk(chunk_key)
-            self.cache.update_used_cache_for_path(chunk_key, chunk.nbytes)  # type: ignore
+    def add_cachables_to_cache_dirty_keys(self):
+        """Adds all the cachables to the cache as dirty keys."""
```

Comment:
Now we add cachable items to dirty keys before a write operation, instead of at the end. This is only done at the time of the first write. This ensures that the objects aren't lost from memory even in case of cache getting filled up as dirty cachables aren't removed from cache even when flushed